keps/sig-node/3815-pod-condition-for-configuration-errors/README.md

```diff
@@ -0,0 +1,1069 @@
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+As a batch user, it is very likely that users experience configuration errors when submitting pods to Kubernetes.  We would like to propose a new condition so that users can determine if a pod has an error in startup.  A new condition will provide a quick way to verify if a pod detected a unrecoverable configuration error.  Many of these configuration errors are fixable via user interactions (ie creating a secret, changing the image name) but without any modifications the pods will stay stuck in the pending state.  
+
+## Motivation
+
+Batch users may not have deep knowledge of Kubernetes.  It is common occurrence to find users submitting Pods with configuration errors (invalid images, wrong image names, missing volumes, missing secrets, wrong volume/secret for config map).  These cases cause the Pod to be stuck in pending and there is code required to remove these pods or user input to remove.
+
+As a batch system developer, we find that higher level controllers have to write code to handle Pending Pod cases.  Since the pods are not considered failed, the higher level controller needs to transition this state to failed in order to enforce retries.  For example, the job controller will run these pods and the state of the Pod will be in Pending.  The job controller does not transition this state to failed even if the pod will never start running.
+
+For CNCF projects, in the Armada project, we implement logic that determines retry ability based on events and statuses.  Events are not standard so we expect issues when updating and we discovered various states that don't have a representative status.  
+
+It would be ideal to have a standard condition for these common configuration errors.  These errors can be fixed via human interaction but they can sometimes be difficult to programatically fix.  A goal of having a standard condition for these configuration errors would be for third-party controllers and/or Kubernetes controllers to start using a single way of detecting fixable configuration errors.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
```

Comment:
Would this cover scenarios like these:

- init container that keeps running and never returns
- postStart hook of a first container never returns and prevents the second container to run
- startup probe never returned `success` and container keep being "started"

Reply:
It could.  I'd have to investigate each of these cases.

Reply:
Sysctl Forbidden is a case that I met before. The kubelet did not add it to the safe sysctl but the apiserver accepted it.

Reply:
> Would this cover scenarios like these:
> 
> * init container that keeps running and never returns
> * init container on the pod with the `restartPolicy=Always` keeps failing and being restarted
> * postStart hook of a first container never returns and prevents the second container to run
> * startup probe never returned `success` and container keep being "started"

I don't think this would cover init containers as I was aiming to test cases where pods are explicitly in pending.  I don't think it doesn't have to support it but this brings up a point that this lifecycle handling could be useful for init containers, ephemeral containers and sidecars.  Currently I was making an assumption that all init containers would have passed successful and then you move onto to main containers.

Reply:
@SergeyKanzhelev  Do we think this should be considered a nongoal or should we consider support for failures in init containers?

Generally these seem to handled a lot better because init containers have a time limit on them and if there are any issues in pulling/running init containers the entire pod lifecycle is blocked.    They are also covered by an `initContainers` condition so it is easy to distinguish that the pod is stuck in the initialized loop (Initialized=False).